### DISCUSSION TITLE
Proper formatting of mathematical expression with Ruff

### DISCUSSION THREAD

**Discussion by cgadal** [astral-sh/ruff#8543] [Q&A] [answered]
First of all, thanks for the great work, ruff is amazing.

However, I have not been able to fully move to ruff since, like black, it does not seem to handle mathematical expressions properly. See [here](https://github.com/psf/black/issues/148) and [here](https://github.com/psf/black/issues/1984).
This is why I moved to `yapf` in the first place, using their knob `ARITHMETIC_PRECEDENCE_INDICATION`.

For example,  `a = np.sqrt(1*2 + 3/5+ (3 +1/5)**(1/(1+3)))` is formatted as:

-  `a = np.sqrt(1 * 2 + 3 / 5 + (3 + 1 / 5) ** (1 / (1 + 3)))` using ruff
-  `a = np.sqrt(1*2 + 3/5 + (3 + 1/5)**(1 / (1+3)))` using yapf with `ARITHMETIC_PRECEDENCE_INDICATION = true`

I can configure VS Code to fix violations on-save using Ruff, then re-format with yapf, using the similar steps indicated in the doc for Black.

However, I'd like to stick to ruff-only ?

How can I achieve this ?

**Comment by MichaReiser** [comment, marked as answer]
Hey @cgadal 

Thanks for your feedback. We received similar feedback around formatting python code in the scientific community in the [Formatter Beta](Discussion). Our main focus for now is compatibility with black instead of innovating a new formatters style, but we'll eventually look into how Ruff can better format complex mathematical expressions (and we'll need the community's feedback to achieve this). But for now, changing the way mathematical expressions isn't planned.

**Reply by ma-sadeghi** [reply]
Is there any updates on this? This would really make ruff even more useful for math-heavy codebases.
